typed/helpers.go

```diff
@@ -56,58 +56,39 @@ func (errs ValidationErrors) Error() string {
 	return strings.Join(messages, "\n")
 }
 
-// errorFormatter makes it easy to keep a list of validation errors. They
-// should all be packed into a single error object before leaving the package
-// boundary, since it's weird to have functions not return a plain error type.
-type errorFormatter struct {
-	path fieldpath.Path
-}
-
-func (ef *errorFormatter) descend(pe fieldpath.PathElement) {
-	ef.path = append(ef.path, pe)
+// Set the given path to all the validation errors.
+func (errs ValidationErrors) WithPath(p string) ValidationErrors {
+	for i := range errs {
+		errs[i].Path = p
+	}
+	return errs
 }
 
-// parent returns the parent, for the purpose of buffer reuse. It's an error to
-// call this if there is no parent.
-func (ef *errorFormatter) parent() errorFormatter {
-	return errorFormatter{
-		path: ef.path[:len(ef.path)-1],
+// Prefix all errors path with the given pathelement. This is useful
+// when unwinding the stack on errors.
+func (errs ValidationErrors) WithPrefix(prefix string) ValidationErrors {
+	for i := range errs {
+		errs[i].Path = prefix + errs[i].Path
 	}
+	return errs
 }
 
-func (ef errorFormatter) errorf(format string, args ...interface{}) ValidationErrors {
+func errorf(format string, args ...interface{}) ValidationErrors {
```

Comment:
Previously, I was very confident the path would always be reported correctly. After this change, we rely on every call site to correctly use WithPrefix or not as needed. Can we change this to capture the path when the error is created? It's fine to convert it to string or whatever.

Reply:
As mentioned before, it didn't work out that way. I can try again.

Reply:
By "path" it doesn't need to be a literal path, a string is fine. I care about where it is captured, not what you store.